src/transaction/atomic_replication_helper.cpp

```diff
@@ -98,24 +96,60 @@ namespace cublog
 	return;
       }
 
-    for (size_t i = 0; i < iterator->second.size (); i++)
-      {
-	iterator->second[i].unfix_page (thread_p);
-      }
-    iterator->second.clear ();
+    iterator->second.unfix_sequence (thread_p);
     m_atomic_sequences_map.erase (iterator);
 
 #if !defined (NDEBUG)
     m_atomic_sequences_vpids_map.erase (tranid);
 #endif
   }
+  /****************************************************************************
+   * atomic_replication_helper::atomic_replication_unit function definitions  *
+   ****************************************************************************/
+  template <typename T>
+  int atomic_replication_helper::atomic_replication_sequence::add_atomic_replication_unit (THREAD_ENTRY *thread_p,
+      log_lsa record_lsa, LOG_RCVINDEX rcvindex, VPID vpid, log_rv_redo_context &redo_context,
+      const log_rv_redo_rec_info<T> &record_info)
+  {
+    m_atomic_replication_unit_vector.emplace_back (record_lsa, vpid, rcvindex);
+    auto iterator = m_atomic_sequence_pages_map.find (vpid);
+    if (iterator == m_atomic_sequence_pages_map.cend ())
+      {
+	int error_code = m_atomic_replication_unit_vector.back ().fix_page (thread_p);
+	if (error_code != NO_ERROR)
+	  {
+	    return error_code;
+	  }
+	m_atomic_sequence_pages_map.emplace (vpid,  m_atomic_replication_unit_vector.back ().get_page_ptr ());
+      }
+    else
+      {
+	m_atomic_replication_unit_vector.back ().set_page_ptr (iterator->second ());
+      }
+    m_atomic_replication_unit_vector.back ().apply_log_redo (thread_p, redo_context, record_info);
+    return NO_ERROR;
+  }
+
+  void atomic_replication_helper::atomic_replication_sequence::unfix_sequence (THREAD_ENTRY *thread_p)
+  {
+    for (size_t i = 0; i < m_atomic_replication_unit_vector.size (); i++)
+      {
+	auto iterator = m_atomic_sequence_pages_map.find (m_atomic_replication_unit_vector[i].m_vpid);
+	if (iterator != m_atomic_sequence_pages_map.end ())
+	  {
+	    m_atomic_replication_unit_vector[i].unfix_page (thread_p);
+	    m_atomic_sequence_pages_map.erase (iterator);
+	  }
+      }
```

Comment:
Unfixes need to happen in the order given by `m_atomic_replication_unit_vector`. Which is what the code does and it is correct.
However, there is no need to check the page and remove it from `m_atomic_sequence_pages_map` because, at the end of the outer iteration, all pgptr's which are also in the `m_atomic_sequence_pages_map` should have been cleared - assuming the two containers are consistent, which they should be.
Thus, the inner part should only be done as a debug-only check at most. If not, a simple vector.clear after all unfixes should be enough.

Reply:
I made this way in order to prevent double unfix on a page, after the VPID is removed from the map if another unit comes with the same VPID it will not perform the unfix.

Reply:
You're right. Mea culpa.